Cross Product for Vector2 and Vector4

AlexRadch opened the pull request:
Close #28731

AlexRadch committed:
System.Numerics Cross Product for Vector2 and Vector4

AlexRadch committed:
Merge remote-tracking branch 'remotes/upstream/main' into Vector2_Vector4_CrossProduct

AlexRadch committed:
Optimize vector operations and update tests

Refactor test methods in `Vector2Tests.cs` and `Vector4Tests.cs` for clarity and correctness.
Enhance the `Cross` methods in `Vector2.cs` and `Vector4.cs` using SIMD techniques for improved performance.

AlexRadch committed:
Refactor Cross method in Vector2.cs for simplicity

Reverted the Cross method to a simpler implementation by removing the use of Vector64 for vector math. The method now directly computes the cross product using the formula `value1.X * value2.Y - value1.Y * value2.X`, prioritizing clarity and performance on xarch architectures.

AlexRadch committed:
Update src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs

Co-authored-by: Huo Yaoyuan <huoyaoyuan@hotmail.com>

AlexRadch committed:
Return [Intrinsic]

AlexRadch committed:
Return New Line

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:
```diff
@@ -323,6 +323,27 @@ public static Vector2 Create(ReadOnlySpan<float> values)
+            // Vector64 is not optimized on xarch. The emulated vector math is much worse
+            // https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgDkBXfGKASzFwG4BYAKDMq0ASgwB2GVsxoBJcW1G52XPn2IBmSgCZyAYXIBvPuWOUN1JOQBmAGwjYMuqBFy4AFADUYYDNG0A3bGsGGCo0ck9vX3IAoJhNAEojE0NeEzTwrx8oVAAeGzsMAD5oqnIAXgzI7JQaHVh7GFcY4NoADTDmkJoATXieVPTjCKzc/Ptiv20K4ehUWvqMRs7NHo7A4JXWvvIAeh3yXAB3bAAHcnbybqTBmeq823HyfAZrBwq/UoAqaM1+6/TiAB2J4vDA0ADiMAwAFFrDBmOJXKR4uQ4CDXhCobD4TBEVQ ...
+
+            return value1.X * value2.Y - value1.Y * value2.X;
```
Yes, it's already a common practice to not use `Vector64`.
The naive math would probably be optimal enough because it's just 3 FP instructions.

EgorBo replied:
it's probably possible to perform multiplications just once on arm64 with vector64, not sure it's worth it

tannergooding replied:
The implementation here should be done what the other paths are doing, which is just using `Vector128`

Functionally this would be:
```csharp
Vector128<float> tmp = left.AsVector128Unsafe() * Vector128.Shuffle(right.AsVector128Unsafe(), Vector128.Create(1, 0, 1, 0));
return (tmp - Vector128.Shuffle(tmp, Vector128.Create(1, 0, 1, 0))).ToScalar();
```

This should be a bit more efficient, particularly on downlevel hardware and the default Crossgen scenario (noting Sharplab is showing AVX compatible codegen): https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvSasAShwB2GXtxYBJGXym5+QkQGYmpBgGEGAbxoNTTbQDMANhGwZ9UCLlwBBABQA1GGAzRdVmAsMNAYvHz8GPgBzAAsMAEoTM2NqMzTQ718oclIADgAea1sMAD4GDHwABwYAXgYAoJYXXDCsnNyAVVVsCxg3eIYAKgzw7LyWAGUYjgtrPui4ppbM6Hau3B6++JDW1fG9WDs+8hC6EJOGOnj44VT002IAdgY3Cuq4EbbxqZm516qdisxrkWAcYEc3Bczgwodd4iwACoQCZgbBWbBQfq3NIAXySpnx5gYRTsDicuAAQp4gf5AsFPhEFglCSl7g9ng0MCwABpDSK8WJcgCaDA+nJYIuGTJ52LMeOoOKAA===

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:
```diff
@@ -385,6 +385,41 @@ public static Vector4 Create(Vector3 vector, float w)
+            Vector128<float> v1 = vector1.AsVector128();
+            Vector128<float> v2 = vector2.AsVector128();
+
+            Vector128<float> m1 = Vector128.Shuffle(v1, Vector128.Create(1, 2, 0, 3)) *
+                Vector128.Shuffle(v2, Vector128.Create(2, 0, 1, 3));
+            Vector128<float> m2 = Vector128.Shuffle(v1, Vector128.Create(2, 0, 1, 3)) *
+                Vector128.Shuffle(v2, Vector128.Create(1, 2, 0, 3));
+
+            return (m1 - m2.WithElement(3, 0)).AsVector4();
```
This should likely be using almost the same code as `Vector3`, not an entirely new algorithm.

AlexRadch replied:
I just did two multiplications and then returned their difference, there is no special algorithm here.
This implementation is simpler and it generates better code in sharplab.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvSasAShwB2GXtxYBJGXym5+QkQGYmpBgGEGAbxoNTZ08W3MkDAGowwGaNr1QIuXJoAU9x84YAbg5OUORodsH+QX5QpACUJuZmxtRJaREx5KQAHAA8AGYANhDYGAB8geQMALyBkaEsAIK4viFZ2QCqqtj5MF5xwqnp5q3Q7QXFpRUBurXRIaRNLfXtXbg9fQM0icOmo6E5EyXlDBgw+AAONRltOSwAygAWHPlFfQFhN2N3rjClfZ9SOE6OFNHE4gwAFRfA7ZB7PV6Fd5AmHtFi/f5eFEghifMFbag7XbEADsqLuAFkOIVZBdCgBPRoAEyZAFFcLJ8JiibszHB9mini83l4PuEBT9YJjseE8eC0DzeXsVnchYjkeKVXCMWcvIDgaD5YqlWdLsaknElvtvAThgBfczbIZpSxiGzW/RuDyaAD6zRQPnq2nmY01MWD9Xi5pSSvJeSKx2mVTmKytWrWG36g1jEvjkxOM2uIdiacyOQzvSzTpzWqOUwY+GTcfhwqRos+ufRUt1+oYOPxUPNw07apFMzDt213b6MtxhttStzdZO+FmzdHbbFzZ1M4Nc4YA8hQ/SI4RY5RnZ3evCs/xg2PFjJXkbDDgDfipZCNuz6QdZmr6SutYMIoJ67i4AG+ygcWHb1NBkYJM6SQxoutYJvWHxFqmzS5lWSHDmh+bTGuxaLDhWp4Q+cbLhUL61KerbvLBZZTn8PY3nuh5USMWotuqooXrxV6znKC68ku6ErmuDH8Vul7Tlie6iYO+GoSxfHnhO3ysZivb9uC96qS6T4vm+q4sAA6rwGCPKySLcDIXiaMC4KftAAZiUkdoAS6VjkO6cFgR4KB+p4gYxPBmRaVAkULIhwwoeJhGJpUWGZG5sJ4TWLE0YEJGRhl7SUUZSQSURpznFctT8rxG5MdFaLCUp84qbGpW1Wem6CepV56fOP68qaVUVZclnWbZ9kwI5zl9p5wxUaSzZUjSvB0oyLLspy3IlQR6l1e2DWSmxAIcX284KjtJ4dYxAmHTpuoied3FmENVGWuREVZb+NB2kAA

Сode from Vector3 generates a negation operation and then replaces the subtraction operation with an addition operation. Otherwise, they are the same.

I can replace my implementation with the one from Vector3.

Which is more appropriate?

tannergooding replied:
> This implementation is simpler and it generates better code in sharplab.

smaller != better

You also have to account for the overhead of simple method having to pull the method arguments in from memory (which will be less typical in real world code), the practical cost of how the code might exist in relation to surrounding code in production, and other considerations.

The implementation that uses `MultiplyAddEstimate` (such that we can opportunistically emits a `FusedMultiplyAdd`) is intentional and should continue being preferred.

The `Vector4` implementation could also be slightly tweaked to avoid the `vinsertps` (by using `vpshufb` instead, where we select `-1` for W instead of `3` in the first multiply, thus setting it to zero). But the 1 cycle savings at the cost of the memory access is likely not worth it.

----------------------------

For the existing Vector3 algorithm, the core of the logic is:
```asm
vpermilps xmm2, xmm1, 0xd2
vxorps xmm2, xmm2, [0x7ffd2f910098]{1to4}
vpermilps xmm3, xmm0, 0xc9
vpermilps xmm1, xmm1, 0xc9
vpermilps xmm0, xmm0, 0xd2
vmulps xmm0, xmm0, xmm1
vfmadd231ps xmm0, xmm3, xmm2
```

It actually should be generating this, but the optimization looks to not be kicking in because its currently only enabled for scalar FMA, not vector FMA:
```asm
vpermilps xmm2, xmm1, 0xd2
vpermilps xmm3, xmm0, 0xc9
vpermilps xmm1, xmm1, 0xc9
vpermilps xmm0, xmm0, 0xd2
vmulps xmm0, xmm0, xmm1
vfnmadd231ps xmm0, xmm3, xmm2
```

Both of these have an approximate throughput of around 7-10 cycles (depending on microarchitecture, with the cost of the xorps is hidden by the latency of the permutes and multiply).

The core of the alternative implementation is:
```
vpermilps xmm2, xmm1, 0xc9
vpermilps xmm3, xmm0, 0xd2
vmulps xmm2, xmm3, xmm2
vpermilps xmm1, xmm1, 0xd2
vpermilps xmm0, xmm0, 0xc9
vmulps xmm0, xmm0, xmm1
vsubps xmm0, xmm2, xmm0
```

This has an approximate throughput of around 10 cycles (with 1 microarchitecture, Broadwell, allowing up to 8 cycles).

So the FMA version has better potential timings and once we ensure the JIT is properly emitting `vfnmadd231ps` will also be smaller. It then has less potential issues with introducing long dependency chains or multiply port contention.

AlexRadch replied:
The only thing that seems to me is that changing the sign has been missed somewhere:

> It actually should be generating this, but the optimization looks to not be kicking in because its currently only enabled for scalar FMA, not vector FMA:
> 
> ```assembly
> vpermilps xmm2, xmm1, 0xd2
> vpermilps xmm3, xmm0, 0xc9
> vpermilps xmm1, xmm1, 0xc9
> vpermilps xmm0, xmm0, 0xd2
> vmulps xmm0, xmm0, xmm1
> vfnmadd231ps xmm0, xmm3, xmm2
> ```
> 
> So the FMA version has better potential timings and once we ensure the JIT is properly emitting `vfnmadd231ps` will also be smaller. It then has less potential issues with introducing long dependency chains or multiply port contention.

tannergooding replied:
It's `vfnmadd231ps` vs `xorps + vfmadd231ps` (note the `n` after the `vf` and before the `madd`, which indicates the negation is happening as part of the instruction).

AlexRadch committed:
Optimize Cross methods in Vector2 and Vector3 using SIMD

Updated the `Cross` method in `Vector2.cs` to utilize SIMD operations with `Vector128`, enhancing performance through hardware acceleration.

The `Cross` method in `Vector3.cs` has been modified to use `AsVector128Unsafe()`, aligning with the performance improvements.

AlexRadch committed:
Improve Cross method implementation in Vector4.cs

Refactor the `Cross` method to enhance performance by utilizing
`Vector128.MultiplyAddEstimate`.

AlexRadch committed:
Refactor Cross method in Vector4.cs for clarity

Removed unnecessary assignment in the Cross method and streamlined the return statement. This change enhances code clarity and efficiency by eliminating redundancy.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:
```diff
@@ -385,6 +384,45 @@ public static Vector4 Create(Vector3 vector, float w)
+                Vector128.Shuffle(v2, shuftleYZXW);
+            m2.SetElementUnsafe(3, 0);
+
+            return (m1 - m2).AsVector4();
```
> Currently JIT can't handle it and will generate much worse code.

The JIT was fixed to handle this in .NET 9 actually.

The issue for `Shuffle` is that there was a flag missing (https://github.com/dotnet/runtime/pull/111303) which blocked the constant from being propagated when there were multiple usages of the local.

AlexRadch replied:
> > Currently JIT can't handle it and will generate much worse code.
> 
> The JIT was fixed to handle this in .NET 9 actually.
> 
> The issue for `Shuffle` is that there was a flag missing (#111303) which blocked the constant from being propagated when there were multiple usages of the local.

Why then does Sharplab show bad JIT asm code if constants are used?

tannergooding replied:
> The issue for Shuffle is that there was a flag missing (https://github.com/dotnet/runtime/pull/111303) which blocked the constant from being propagated when there were multiple usages of the local.

It's an issue specific to shuffle because the `local` is used in 2 places. If you check a local that is only used once, you'll see good codegen

See that it's only M3 that has bad codegen here: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvSasAShwB2GXtxYBJGXym5+QmiIDMTUgwDCNAN40GZpjuZIGANRhgM0cqQAcAHgBmAGwjYMAPgYAWXIACjsHJ1dPHz9AgDdyNFt7Ryhnd29fAIZ40gBKU3MTanMylMj06N4ZQNwACw4PDC8YAE0ALQANAHUGAF4KtIyWfVg/GFCkhlJkumStfOFS8rMI4erahgamlphutr7B9aiXUfGMSdmGeYZpxeWi1eIAdiHTlgBlRo9vScTkjtmq1Or1CitVuUANTvKpnb5NP6hPKAn57A49JZPBgAX2x2OIlnI1hOcJi2UCQVI4VSp3JcVy01JGXpOTy4LKJUhsJZNRyQL2oKOPNc5xgEymyWutwemgh5VeIvhPyRAO2aJB3Ux2O5DBhzNFCN+rWR1wNZzG4sukpmcwW+Sx8rx8oJRJJtLJWQZQS0NMqLK9bKZHoDsTZBWxXMh5rcfLqGvaWoGSrFEum0vtj3lZUV5q+KpNaoFmrBOu5+pDhoL/2uxcTYOWZWdOKAA==

The PR I linked is fixing the issue

AlexRadch replied:
This is great PR! I would also add at least one test to ensure that this optimization is not accidentally broken in the future.

Should I leave code that has optimal code generation before .NET 9?

hez2010 replied:
The codegen has been fixed: https://godbolt.org/z/3qabvjvd6
You can keep the original code as is, this new API won't be backported to .NET 9.

tannergooding replied:
> Should I leave code that has optimal code generation before .NET 9?

I would recommend continuing to use the older pattern, yes. That is the one that explicitly inlines the constant.

Not only is that code more portable, in case someone tries to copy the implementation for downlevel usage, but it also ensures that the code is better behaved on other runtimes like Mono, and additionally reduces the amount of work that RyuJIT needs to do (as it doesn't need to try and propagate the constants back to the shuffle).

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:
```diff
@@ -368,8 +368,8 @@ public static Vector3 Cross(Vector3 vector1, Vector3 vector2)
 
-            Vector128<float> v1 = vector1.AsVector128();
-            Vector128<float> v2 = vector2.AsVector128();
+            Vector128<float> v1 = vector1.AsVector128Unsafe();
+            Vector128<float> v2 = vector2.AsVector128Unsafe();
```
If you're going to do this, you should modify the `Shuffle` below to use `-1`, rather than `3`, for selecting `W`

It avoids a potential pessimization on some hardware if `W` is `NaN` or similar and is "free".

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:
```diff
@@ -323,6 +323,28 @@ public static Vector2 Create(ReadOnlySpan<float> values)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static float Cross(Vector2 value1, Vector2 value2)
+        {
+            //return value1.X * value2.Y - value1.Y * value2.X;
+
+            Vector128<float> mul = value1.AsVector128Unsafe() *
+                Vector128.Shuffle(value2.AsVector128Unsafe(), Vector128.Create(1, 0, 1, 0));
+            return (mul - Vector128.Shuffle(mul, Vector128.Create(1, 0, 1, 0))).ToScalar();
```
Similar here. It is preferable to use `AsVector128` or to guarantee the zeroing of the upper elements by using `-1` as the index in the shuffle. It helps avoid pessimization and should be naturally part of the underlying codegen without inserting additional instructions

AlexRadch committed:
Optimize Cross product calculations in Vector2 and Vector3

Refactored the `Cross` method in `Vector2.cs` to enhance performance by using vector shuffling and subtraction with SIMD operations. Updated the `Cross` method in `Vector3.cs` to correct shuffling indices for accurate component multiplication, also leveraging SIMD for improved efficiency.
